### DISCUSSION TITLE
RTMP Publish External Authentication Empty POST Data

### DISCUSSION THREAD

**Discussion by chirenonhive** [bluenviron/mediamtx#3006] [Questions]
### Question

I've been using nginx rtmp as a streaming server and authenticating users using an external server.

Here's how it currently works:
User sets `rtmp://serverurl/live` as the server in OBS
and set the stream key as `USERNAME?key=STREAMKEYHERE`

When the user starts streaming, it's authenticated by sending a get request (`http://serverurl/on_publish?key=STREAMKEYHERE`) to my authentication url. The server return a 200 response if the key exists in the database and allow the stream to be published.

Now I want to achieve something similar with MediaMTX. I've added handling for the POST request that MediaMTX sends to the external authentication url, but it's always sending and empty post request. I've tried both setting the stream key in OBS to `?user=USERNAME&pass=STREAMKEYHERE` and ticking the "Use authentication" box and inputting the user and password there. But the post request received by my authentication server is always empty.

Seeing as the documentation says it wont request credentials until it receives a 401 response, I do return a 401 response when user/pass is empty, but then it always returns 401.

What am I doing wrong?

Is it possible to achieve what I want with MediaMTX?

**Comment by aler9** [comment]
Hello,

> Seeing as the documentation says it wont request credentials until it receives a 401 response, I do return a 401 response when user/pass is empty, but then it always returns 401.

This passage refers to the RTSP protocol, not the RTMP protocol.

> I've tried both setting the stream key in OBS to ?user=USERNAME&pass=STREAMKEYHERE and ticking the "Use authentication" box and inputting the user and password there. But the post request received by my authentication server is always empty.

None of this steps is the one described in the README, in the [OBS Studio](https://github.com/bluenviron/mediamtx?tab=readme-ov-file#obs-studio) section, that tells:

> If credentials are in use, use the following parameters:
>    Service: Custom...
>    Server: rtmp://localhost
>    Stream key: mystream?user=myuser&pass=mypass

Therefore, the path name is missing from the stream key and you don't have to tick the "Use authentication" box. Check the README if you still have doubts.

[a comment by an automated account is not shown]
